### DISCUSSION TITLE
Bucket4j and Javalin

### DISCUSSION THREAD

**Discussion by tipsy** [vladimir-bukhtoyarov/bucket4j#243] [Ideas]
Hi @vladimir-bukhtoyarov - I run the Javalin open source project (https://github.com/tipsy/javalin), and when people ask for rate-limiting, I usually refer them to bucket4j. Do you think you could add an example using Javalin, and/or write an official bucket4j Javalin plugin? I would of course be happy to help with the Javalin related parts of the example/plugin, but I don't feel familiar enough with bucket4j to provide an official example to my users.

**Comment by vladimir-bukhtoyarov** [comment]
Hello @tipsy,

I will create pull request with usage example to Javalin repo at weekend. It will contain functionality that similar to already written by yourself in the [RateLimitUtil.kt](https://github.com/tipsy/javalin/blob/95ab1d2a75ed6af3b6863c9dd8af6ab29acb59ac/javalin/src/main/java/io/javalin/http/util/RateLimitUtil.kt).

I have several thoughts about idea to implement Javalin plugin:
* It should be a part of Javalin repository, not a part of Bucket4j repository, because library must not depend from framework. For example Springframework always depends from Hibernate and never vise-versa.
* Example of usage will be pretty easy to understand, I hope that creating plugin by yourself will be an easy exercise.
* Typically frameworks like Javalin provides pluggable API for rate-limiting, it makes possible for end user to choose concrete rate-limiting library, so may be it will be good to bring rate-limiting related abstractions to your framework.

Small off-top regarding to source code of [RateLimitUtil.kt](https://github.com/tipsy/javalin/blob/95ab1d2a75ed6af3b6863c9dd8af6ab29acb59ac/javalin/src/main/java/io/javalin/http/util/RateLimitUtil.kt) - storing limiters inside concurrent hashmap without expiration can be a cause of OOM because hashmap will continuously increase each time when new key found, and will not decrease when old key are not used for long time. In the my example I will use Bucket4j integration with Caffeine in order to bring expiration and maximum size of collection with buckets.

**Reply by tipsy** [reply]
Thanks for your notes on the internal rate-limiter @vladimir-bukhtoyarov, I've created an issue for fixing that. The class is named `NaiveRateLimit` to discourage production usage, but it's of course better to avoid memoryleaks altogether.

I'm happy to host an example in the Javalin org, let me prepare a repository where you have write-access :)

**Reply by tipsy** [reply]
I've created https://github.com/javalin/javalin-bucket4j - I'll push a "Hello World" app shortly

**Reply by vladimir-bukhtoyarov** [reply]
Got it.

**Reply by vladimir-bukhtoyarov** [reply]
@tipsy I have pushed simple usage example https://github.com/javalin/javalin-bucket4j/commit/d9db6c61c22cf26d4f0274962ea3d0806397ff1b

More examples you can find in the reference https://bucket4j.com/7.3.0/toc.html#quick-start-examples

**Reply by tipsy** [reply]
Thank you very much @vladimir-bukhtoyarov !

**Reply by vladimir-bukhtoyarov** [reply]
Also, I highly recommend to look to the https://github.com/MarcGiffing/bucket4j-spring-boot-starter as a good example of how to provide for end users the zero-code solution  on top of Bucket4j.

**Reply by tipsy** [reply]
Thanks for the tip @vladimir-bukhtoyarov - I'll make an issue on my tracker and see if anyone wants to create a plugin based on the starter project.
